File: java/findClosestValueInBST.java
class FindClosestValueInBst {

  public static int findClosestValueInBst(BST tree, int target) {
    return searchTree(tree, target, tree.value);
  }

  public static int searchTree(BST tree, int target, int closestVal) {
    int currentDifference = Math.abs(closestVal - target);
    int curVal = tree.value;
    int newDifference = Math.abs(curVal - target);

    if (newDifference < currentDifference) {
      closestVal = curVal;
    }

    if (curVal > target && tree.left != null) {
      closestVal = searchTree(tree.left, target, closestVal);
    } else if (curVal < target && tree.right != null) {
      closestVal = searchTree(tree.right, target, closestVal);
    }

    return closestVal;

  }

  static class BST {
    public int value;
    public BST left;
    public BST right;

    public BST(int value) {
      this.value = value;
    }
  }
}

File: javaScript/findClosestValueInBST.js
function findClosestValueInBst(tree, target) {
  let closest = tree.value;
  let closestDiff = Math.abs(tree.value - target);

  function searchBST(node) {
    let val = node.value;
    let diff = Math.abs(val - target);

    if (diff < closestDiff) {
      closestDiff = diff;
      closest = val;
    }

    if (val < target) {
      if (node.right !== null) {
        searchBST(node.right);
      }
    } else if (val > target) {
      if (node.left !== null) {
        searchBST(node.left);
      }
    } else if (val === target) {
      closest = val;
      return;
    }
  }

  searchBST(tree);
  return closest;
}
